### PR TITLE
[JENKINS-69890] Prevent deadlock on websocket agents

### PR DESCRIPTION
**Pull request by amuniz**
More information in [JENKINS-69890](https://issues.jenkins.io/browse/JENKINS-69890)

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/hudson/remoting/Engine.java [hunk 1]
```diff
@@ -625,7 +625,9 @@ private void onMessage(ByteBuffer message) {
                     @Override
                     public void onClose(Session session, CloseReason closeReason) {
                         LOGGER.fine(() -> "onClose: " + closeReason);
-                        transport.terminate(new ChannelClosedException(ch.get(), null));
+                        // making this call async to avoid potential deadlocks when some thread is holding a lock on the
+                        // channel object while this thread is trying to acquire it to call Transport#terminate
+                        ch.get().executor.submit(() -> transport.terminate(new ChannelClosedException(ch.get(), null)));
```
Comment (Vlatombe) [inline src/main/java/hudson/remoting/Engine.java:630]: I suspect a similar change would be needed in `onError`

**Review by jglick — APPROVED** [review]
Needs at least a manual sanity check. You can also check in core `WebSocketAgentsTest` and `JNLPLauncherRealTest.webSocket`.

**Comment by basil** [comment]
I wonder if this could be one cause of occasional flakes in `JnlpSlaveRestarterInstallerTest#webSocketReconnection` like https://ci.jenkins.io/job/Core/job/jenkins/job/PR-7283/1/testReport/junit/jenkins.slaves.restarter/JnlpSlaveRestarterInstallerTest/Windows_jdk11___Windows_Build___Test___webSocketReconnection/ — I've never seen that test flake with TCP reconnection.

File: src/main/java/hudson/remoting/Engine.java (hunk 1 above)
Comment (aneveux) [reply src/main/java/hudson/remoting/Engine.java:630]: :+1: 

I just added it as well.

**Comment by aneveux** [comment]
@jglick Good idea!

On top of running the automated tests of that project, I also built it and used it in core in order to run the tests you mentioned, and they went fine.

File: src/main/java/hudson/remoting/Engine.java [hunk 2]
```diff
@@ -633,7 +633,9 @@ public void onClose(Session session, CloseReason closeReason) {
                     public void onError(Session session, Throwable x) {
                         // TODO or would events.error(x) be better?
                         LOGGER.log(Level.FINE, null, x);
-                        transport.terminate(new ChannelClosedException(ch.get(), x));
+                        // making this call async to avoid potential deadlocks when some thread is holding a lock on the
+                        // channel object while this thread is trying to acquire it to call Transport#terminate
```
Comment (jglick) [inline src/main/java/hudson/remoting/Engine.java:637]:
```suggestion
                        // as above
```

**Review by Vlatombe — APPROVED** [review]

**Review by jeromepochat — APPROVED** [review]

**Comment by amuniz** [comment]
Two spotbugs issues.

**Review by basil — APPROVED** [review]

**Pull request merged**